File: android/aisupport-sdk/src/main/java/io/polydev/aisupport/APIClient.kt
package io.polydev.aisupport

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * AISUP API Client for Android
 * Handles REST API communication with AISup backend
 */
class AISUPAPIClient(private val config: AISUPConfig) {
    
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        isLenient = true
    }
    
    var chatId: String? = null
        private set
    
    fun setChatId(id: String) {
        chatId = id
    }
    
    /**
     * Initialize a new chat session
     */
    suspend fun initialize(): InitResponse = withContext(Dispatchers.IO) {
        val body = buildString {
            append("{")
            append("\"userName\":\"${config.userName}\"")
            chatId?.let { append(",\"chatId\":\"$it\"") }
            append("}")
        }
        
        val request = Request.Builder()
            .url("${config.apiUrl}/api/integration/init")
            .addHeader("Content-Type", "application/json")
            .addHeader("X-API-Key", config.apiKey)
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()
        
        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw AISUPException.ServerError("HTTP ${response.code}")
        }
        
        val responseBody = response.body?.string() 
            ?: throw AISUPException.InvalidResponse()
        
        val result = json.decodeFromString<InitResponse>(responseBody)
        chatId = result.chatId
        result
    }
    
    /**
     * Send a text message
     */
    suspend fun sendMessage(
        content: String, 
        attachments: List<Attachment>? = null
    ): SendMessageResponse = withContext(Dispatchers.IO) {
        val currentChatId = chatId ?: throw AISUPException.NotInitialized()
        
        val body = buildString {
            append("{")
            append("\"chatId\":\"$currentChatId\",")
            append("\"content\":\"${content.replace("\"", "\\\"")}\"")
            attachments?.let {
                append(",\"attachments\":${json.encodeToString(it)}")
            }
            append("}")
        }
        
        val request = Request.Builder()
            .url("${config.apiUrl}/api/integration/send-message")
            .addHeader("Content-Type", "application/json")
            .addHeader("X-API-Key", config.apiKey)
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()
        
        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw AISUPException.ServerError("HTTP ${response.code}")
        }
        
        val responseBody = response.body?.string() 
            ?: throw AISUPException.InvalidResponse()
        
        json.decodeFromString<SendMessageResponse>(responseBody)
    }
    
    /**
     * Get message history
     */
    suspend fun getMessages(
        limit: Int = 50, 
        before: String? = null
    ): MessagesResponse = withContext(Dispatchers.IO) {
        val currentChatId = chatId ?: throw AISUPException.NotInitialized()
        
        val urlBuilder = StringBuilder("${config.apiUrl}/api/integration/messages")
        urlBuilder.append("?chatId=$currentChatId")
        urlBuilder.append("&limit=$limit")
        before?.let { urlBuilder.append("&before=$it") }
        
        val request = Request.Builder()
            .url(urlBuilder.toString())
            .addHeader("X-API-Key", config.apiKey)
            .get()
            .build()
        
        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw AISUPException.ServerError("HTTP ${response.code}")
        }
        
        val responseBody = response.body?.string() 
            ?: throw AISUPException.InvalidResponse()
        
        json.decodeFromString<MessagesResponse>(responseBody)
    }
    
    /**
     * Upload a file
     */
    suspend fun uploadFile(
        file: File, 
        mimeType: String
    ): UploadResponse = withContext(Dispatchers.IO) {
        val currentChatId = chatId ?: throw AISUPException.NotInitialized()
        
        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart("chatId", currentChatId)
            .addFormDataPart(
                "file",
                file.name,
                file.readBytes().toRequestBody(mimeType.toMediaType())
            )
            .build()
        
        val request = Request.Builder()
            .url("${config.apiUrl}/api/integration/upload")
            .addHeader("X-API-Key", config.apiKey)
            .post(requestBody)
            .build()
        
        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            throw AISUPException.ServerError("HTTP ${response.code}")
        }
        
        val responseBody = response.body?.string() 
            ?: throw AISUPException.InvalidResponse()
        
        json.decodeFromString<UploadResponse>(responseBody)
    }
    
    /**
     * Mark messages as read
     */
    suspend fun markAsRead(): Unit = withContext(Dispatchers.IO) {
        val currentChatId = chatId ?: return@withContext
        
        val body = "{\"chatId\":\"$currentChatId\"}"
        
        val request = Request.Builder()
            .url("${config.apiUrl}/api/integration/mark-read")
            .addHeader("Content-Type", "application/json")
            .addHeader("X-API-Key", config.apiKey)
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()
        
        client.newCall(request).execute()
    }
}

File: src/api-client.ts
/**
 * AISup API Client
 * Standalone API client for integration with AISup backend
 * Can be used without the UI widget
 */

export interface AISupportConfig {
  apiKey: string;
  apiUrl: string;
  wsUrl?: string;
  userName?: string;
}

export interface Message {
  _id: string;
  chat: string;
  content: string;
  role: 'user' | 'bot' | 'admin';
  type: 'text' | 'photo' | 'file' | 'audio' | 'video' | 'video_note' | 'voice';
  caption?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Attachment {
  type: 'image' | 'video' | 'file';
  url: string;
  name: string;
  size?: number;
  mimeType?: string;
}

export interface Chat {
  _id: string;
  chatId: string;
  platform: string;
  status: 'active' | 'closed' | 'pending';
  mode: 'bot' | 'operator';
  userName?: string;
  createdAt: string;
  updatedAt: string;
}

export interface InitResponse {
  response: string;
  message: string;
  data: {
    chatId: string;
    startMessage: string;
  };
}

export interface SendMessageResponse {
  response: string;
  message: string;
}

export interface MessagesResponse {
  response: string;
  message: string;
  data: Message[];
}

export interface UploadResponse {
  response: string;
  message: string;
  data: {
    url: string;
    type: string;
    originalName: string;
    size: number;
    mimetype: string;
    messageId: string;
  };
}

export class AISupportAPIClient {
  private config: AISupportConfig;
  private chatId: string | null = null;

  constructor(config: AISupportConfig) {
    this.config = {
      ...config,
      wsUrl: config.wsUrl || config.apiUrl,
      userName: config.userName || 'Guest',
    };
  }

  /**
   * Get the current chat ID
   */
  getChatId(): string | null {
    return this.chatId;
  }

  /**
   * Set chat ID (useful when restoring from storage)
   */
  setChatId(chatId: string): void {
    this.chatId = chatId;
  }

  /**
   * Get configuration
   */
  getConfig(): AISupportConfig {
    return { ...this.config };
  }

  /**
   * Initialize a new chat session
   */
  async init(): Promise<InitResponse> {
    const clientChatId = this.chatId || crypto.randomUUID();
    
    const response = await fetch(`${this.config.apiUrl}/api/integration/init`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.config.apiKey,
      },
      body: JSON.stringify({
        chatId: clientChatId,
        chatNickname: this.config.userName,
      }),
    });

    if (!response.ok) {
      throw new Error(`Init failed: ${response.status} ${response.statusText}`);
    }

    const data: InitResponse = await response.json();
    this.chatId = data.data.chatId;
    return data;
  }

  /**
   * Send a text message
   */
  async sendMessage(content: string, attachments?: Attachment[]): Promise<SendMessageResponse> {
    if (!this.chatId) {
      throw new Error('Chat not initialized. Call init() first.');
    }

    const response = await fetch(`${this.config.apiUrl}/api/integration/send-message`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.config.apiKey,
      },
      body: JSON.stringify({
        chatId: this.chatId,
        messageText: content,
      }),
    });

    if (!response.ok) {
      throw new Error(`Send message failed: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get message history
   */
  async getMessages(limit = 50, before?: string): Promise<MessagesResponse> {
    if (!this.chatId) {
      throw new Error('Chat not initialized. Call init() first.');
    }

    const response = await fetch(`${this.config.apiUrl}/api/integration/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.config.apiKey,
      },
      body: JSON.stringify({
        chatId: this.chatId,
      }),
    });

    if (!response.ok) {
      throw new Error(`Get messages failed: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Upload a file
   */
  async uploadFile(file: File): Promise<UploadResponse> {
    if (!this.chatId) {
      throw new Error('Chat not initialized. Call init() first.');
    }

    const formData = new FormData();
    formData.append('file', file);
    formData.append('chatId', this.chatId);

    const response = await fetch(`${this.config.apiUrl}/api/integration/upload`, {
      method: 'POST',
      headers: {
        'X-API-Key': this.config.apiKey,
      },
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`Upload failed: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Mark messages as read
   */
  async markAsRead(): Promise<void> {
    if (!this.chatId) {
      return;
    }

    await fetch(`${this.config.apiUrl}/api/integration/mark-read`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.config.apiKey,
      },
      body: JSON.stringify({
        chatId: this.chatId,
      }),
    });
  }
}

export default AISupportAPIClient;
